src/Symfony/Component/PropertyInfo/Util/LegacyTypeConverter.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\PropertyInfo\Util;
+
+use Symfony\Component\PropertyInfo\Type as LegacyType;
+use Symfony\Component\TypeInfo\Type;
+
+/**
+ * @internal
+ */
+class LegacyTypeConverter
+{
+    /**
+     * @param LegacyType[]|null $legacyTypes
+     */
+    public static function toTypeInfoType(?array $legacyTypes): ?Type
+    {
+        if (null === $legacyTypes) {
+            return null;
+        }
+
+        $types = [];
+
+        foreach ($legacyTypes as $legacyType) {
+            $typeInfoType = match ($legacyType->getBuiltinType()) {
+                LegacyType::BUILTIN_TYPE_ARRAY => Type::array(),
```

Comment:
this should use `getCollectionKeyTypes` and `getCollectionValueTypes` to avoid loosing information available in the legacy type (same for iterable).